File: python/백준/자료구조/덱, 큐/5430 AC.py
import sys
from collections import deque

input = sys.stdin.readline

T = int(input())  # 테스트 케이스 수 입력

for _ in range(T):
    # 입력 처리
    p = input().strip()  # 명령어
    n = int(input())  # 배열 크기
    arr = input().strip()[1:-1]  # 배열 형태의 문자열
    
    # 배열 초기화
    if n == 0:
        queue = deque()  # 빈 배열일 경우
    else:
        queue = deque(arr.split(','))  # 배열로 변환
    
    reverse_flag = False  # 뒤집기 여부를 추적
    error_flag = False  # 에러 발생 여부 추적
    
    # 명령어 처리
    for cmd in p:
        if cmd == 'R':
            reverse_flag = not reverse_flag  # 뒤집기 상태 변경
        elif cmd == 'D':
            if queue:
                if reverse_flag:
                    queue.pop()  # 뒤집힌 상태라면 뒤에서 제거
                else:
                    queue.popleft()  # 정상 상태라면 앞에서 제거
            else:
                error_flag = True  # 에러 발생
                break
    
    # 결과 출력
    if error_flag:
        print("error")
    else:
        if reverse_flag:
            queue = reversed(queue)  # 결과적으로 뒤집어서 출력
        print('[' + ",".join(queue) + ']')

File: cpp/백준/자료구조/덱, 큐/5430 AC.cpp
#include <iostream>
#include <deque>
#include <string>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        string p;
        int n;
        string arr;
        cin >> p >> n >> arr;

        deque<string> queue;
        if (n > 0) {
            string body = arr.substr(1, arr.size() - 2);
            string cur;
            for (char c : body) {
                if (c == ',') {
                    queue.push_back(cur);
                    cur.clear();
                } else {
                    cur += c;
                }
            }
            if (!cur.empty()) queue.push_back(cur);
        }

        bool reverse_flag = false;
        bool error_flag = false;

        for (char cmd : p) {
            if (cmd == 'R') {
                reverse_flag = !reverse_flag;
            } else if (cmd == 'D') {
                if (!queue.empty()) {
                    if (reverse_flag) queue.pop_back();
                    else queue.pop_front();
                } else {
                    error_flag = true;
                    break;
                }
            }
        }

        if (error_flag) {
            cout << "error\n";
        } else {
            cout << "[";
            if (reverse_flag) {
                for (int i = (int)queue.size() - 1; i >= 0; --i) {
                    cout << queue[i];
                    if (i > 0) cout << ",";
                }
            } else {
                for (int i = 0; i < (int)queue.size(); ++i) {
                    cout << queue[i];
                    if (i + 1 < (int)queue.size()) cout << ",";
                }
            }
            cout << "]\n";
        }
    }
    return 0;
}
